enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,339 @@
+
+    apiVIPs:
+    - 192.168.111.5
+    - f00::5
+    ingressVIPs:
+    - 192.168.111.4
+    - f00::4
+```
+This is essentially a variation on the previous option where the two separate
+fields per VIP are combined into a list field with a length of up to 2. In
+this case we'd need some logic to determine whether the VIPs in the list are
+v4 or v6, but I expect that would only be a little more complicated than the
+other options. This has the advantage of being extensible if we ever wanted
+to add more addresses, but I admit to having a hard time coming up with a use
+case where more than an ipv4 and ipv6 address would be needed.
+
+### openshift/api
+
+We will need to make changes to the
+[baremetal platformStatus](https://github.com/openshift/api/blob/354aa98a475c1fcd60b41aaee735da95d7318b42/config/v1/0000_10_config-operator_01_infrastructure.crd.yaml#L311)
+to add new fields for the new VIPs. Unlike the install-config layout, I think
+it is clearly better to add discrete new fields for this feature. That way,
+backend code in machine-config-operator and baremetal-runtimecfg will not
+need to detect whether the values in the old fields are from a new (where
+the field will have to be v4) or old (where the existing field can be either
+v4 or v6) configuration.
+
+Ideally we would transition existing values to the new fields on upgrade.
+I'm unsure whether a mechanism exists for us to do that, however. If not, then
+some logic (probably in MCO) will be needed to map values from the old fields
+to the appropriate ip version.
+
+The new structure of platformStatus would look like this:
+
+```
+baremetal:
+  description: BareMetal contains settings specific to the BareMetal platform.
+  type: object
+  properties:
+    # Deprecated
+    apiServerInternalIP:
+      description: apiServerInternalIP is an IP address...
+      type: string
+    apiServerInternalIPv4:
+      description: apiServerInternalIPv4 is an IP address...
+      type: string
+    apiServerInternalIPv6:
+      description: apiServerInternalIPv6 is an IP address...
+      type: string
+    # Deprecated
+    ingressIP:
+      description: ingressIP is an external IP...
+      type: string
+    ingressIPv4:
+      description: ingressIPv4 is an external IP...
+      type: string
+    ingressIPv6:
+      description: ingressIPv6 is an external IP...
+      type: string
+    nodeDNSIP:
+    ...
+```
+
+### machine-config-operator
+
+We will need to wire in the new VIPs to the manifests and configuration
+templates. There should be conditionals around the v4 and v6 sections
+so only the appropriate ones are enabled based on the provided VIPs.
+
+I believe haproxy already listens on both ipv4 and ipv6 so there should be no
+changes needed in that configuration. We'll need to verify that as part of
+this implementation though. Note that as of this writing, the apiserver
+does not support dual stack anyway so all traffic through haproxy will end
+up on one or the other ip version.
```

Comment:
FTR, the apiserver situation is slightly more complicated than this: it doesn't support having the `kubernetes.default` Service be dual-stack, and it doesn't have any way of recognizing / talking about dual-stack IPs, but it _can_ take advantage of Linux's support for dual-stack listening sockets. Currently cluster-kube-apiserver-operator explicitly passes `bindNetwork: tcp4` or `bindNetwork: tcp6` in the kube-apiserver config to prevent that, but if we changed it to pass `bindNetwork: tcp` instead, then we'd get an apiserver that accepted connections on both the IPv4 and IPv6 IPs of its node (even though it would only create `Endpoints` for one family or the other).

Reply:
Yeah, this was a bit hand-wavy because I couldn't remember exactly what you said was missing from dual stack in the apiserver. I'll rewrite this to be more accurate.